### PR TITLE
Ability to download XML and TSV report.

### PR DESCRIPTION
**Pull request by vivekiitkgp**
Signed-off-by: Vivek Rai <vivekraiiitkgp@gmail.com>

### CODE REVIEW

**Comment by vivekiitkgp** [comment]
Will need to rework the test suite a bit.

**Comment by yannickwurm** [comment]
Thanks - Heads up that Travis build is failing. 

Can this be finalised so it can be deployed to `--pre` gem later today so we can announce the whole prerelease to mailing list before the weekend?

**Comment by vivekiitkgp** [comment]
@yannickwurm Do you like this?

**Comment by yeban** [comment]
@vivekiitkgp I think we should provide two links in the sidebar:

    TSV Report: [standard] [full]

And detail the columns [standard] will contain in a BS tooltip.

If you want to provide TSV customisation, you could add a 3rd link (if space allows) - [custom] which opens the modal in its current form sans the checkboxes. I'm up for custom TSV, but I would like to click on the name of columns rather than type them out in an input box.

**Comment by vivekiitkgp** [comment]
@yannickwurm I would also like to have few words from you before we modify the current implementation. An early feedback will reduce total invested time.

**Comment by yannickwurm** [comment]
Hey @vivekiitkgp - give me a screenshot please. 
 1. Naming:  ~~TSV Report:~~ -> "Tab-delimited table:" (most humans haven't heard of TSV). 
 2. We do not want/need customisation - noone will ever use it; and it adds implementation/maintenance overhead (options can differ btw blast versions...)

**Comment by vivekiitkgp** [comment]
@yannickwurm Here are possible use case scenarios and relevant screenshots.

screenshot-localhost 4567 2015-02-01 08-05-36
screenshot-localhost 4567 2015-02-01 08-05-05
screenshot-localhost 4567 2015-02-01 08-04-29

**Comment by yannickwurm** [comment]
Whoa - this is waaaaay overkill.=20

No modals. No extra unnecessary clicking.



Just do the simplest thing possible:=20
Just replace "TSV report (coming soon)"  with "Tab delimited report =
[Standard] [Full]"





> On 1 Feb 2015, at 02:36, Vivek Rai <notifications@github.com> wrote:
>=20
> @yannickwurm Here are possible use case scenarios and relevant =
screenshots.

**Comment by yeban** [comment]
Please rebase on top of latest master. A lot has changed in the last 3 commits and there will be merge conflicts.

File: spec/ss_sample_blast_no_hits.asn [hunk 1]
```diff
@@ -0,0 +1,130 @@
+Blast4-archive ::= {
```
Comment (yeban) [inline spec/ss_sample_blast_no_hits.asn]: Archive format (11) is binary, I believe. How come using text file for test suite?

Comment (vivekiitkgp) [inline spec/ss_sample_blast_no_hits.asn]: It (outfmt 11) is a readable text file.

File: lib/sequenceserver.rb [hunk 2]
```diff
@@ -16,6 +16,21 @@ module SequenceServer
   # Use a fixed minimum version of BLAST+
   MINIMUM_BLAST_VERSION           = '2.2.30+'
 
+  OUTFMT = {
```
Comment (yeban) [inline lib/sequenceserver.rb]: This should be in BLAST.

Comment (yeban) [inline lib/sequenceserver.rb]: Together with the reference link [1] below.

File: lib/sequenceserver.rb [hunk 3]
```diff
@@ -373,6 +388,26 @@ def tuple?(data)
...
+      rfile  = File.join(Dir.tmpdir, params[:report])
+
+      file_name = "sequenceserver_report_#{Time.now.strftime('%H%M')}.#{format}"
+      file = Tempfile.new file_name
+
+      case format
+      when 'xml'
+        file.write(BLAST.blast_formatter(rfile, OUTFMT['xml']))
+      when 'tsv'
+        file.write(BLAST.blast_formatter(rfile, OUTFMT['tsv'], specifiers))
```
Comment (yeban) [inline lib/sequenceserver.rb]: Commented TSV could be more helpful.

File: lib/sequenceserver/blast.rb [hunk 4]
```diff
@@ -104,6 +103,10 @@ def pre_process(params)
         params[:sequence].strip! unless params[:sequence].nil?
       end
 
+      def blast_formatter(rfile_path, ofmt, specifiers = nil)
+        `blast_formatter -archive '#{rfile_path}' -outfmt '#{ofmt} #{specifiers}'`
```
Comment (yeban) [inline lib/sequenceserver/blast.rb]: `specifiers` is technically a user input. Thus should be verified, and used only if applicable. That is, only for TSV, and only after checking its validity.

File: lib/sequenceserver.rb [hunk 5]
```diff
@@ -373,6 +388,26 @@ def tuple?(data)
       end
     end
 
+    get '/get_report/' do
+      format = params[:format]
+      specifiers = params[:specifiers]
+
+      rfile  = File.join(Dir.tmpdir, params[:report])
```
Comment (yeban) [inline lib/sequenceserver.rb]: Verify it's existence before using it.

File: spec/ss_sample_blast_no_hits.asn (hunk 1 above)
Comment (yeban) [inline spec/ss_sample_blast_no_hits.asn]: How?

File: lib/sequenceserver/blast.rb (hunk 4 above)
Comment (vivekiitkgp) [inline lib/sequenceserver/blast.rb]: A good thing about specifiers here is that wrong values are happily ignored. So if I pass `specifiers  = "qframe random none"`, then output will contain only of `qframe`.

**Comment by yeban** [comment]
Regarding failing test suite, do note that rubocop is a part of the test suite now.

File: spec/ss_sample_blast_no_hits.asn (hunk 1 above)
Comment (vivekiitkgp) [inline spec/ss_sample_blast_no_hits.asn]: Binary ASN.1 is `-outfmt 9`

File: lib/sequenceserver/blast.rb (hunk 4 above)
Comment (yeban) [inline lib/sequenceserver/blast.rb]: Fair enough. Add a note in the source code though.

What's the behaviour with XML?

Comment (vivekiitkgp) [inline lib/sequenceserver/blast.rb]: Doesn't affect it.

Comment (yeban) [inline lib/sequenceserver/blast.rb]: Fine then. Let BLAST deal with incorrect values (if at all). Do make a note of the behaviour.

**Comment by yeban** [comment]
Keep the controllers as thin as possible. Like, `post '/'`. I think all the logic of `get_report/` can be moved to `BLAST.blast_formatter`.

**Comment by yeban** [comment]
And should call the method `format` instead of `blast_formatter`.

File: lib/sequenceserver.rb [hunk 6]
```diff
@@ -373,6 +388,26 @@ def tuple?(data)
...
+
+      file_name = "sequenceserver_report_#{Time.now.strftime('%H%M')}.#{format}"
+      file = Tempfile.new file_name
+
+      case format
+      when 'xml'
+        file.write(BLAST.blast_formatter(rfile, OUTFMT['xml']))
```
Comment (yeban) [inline lib/sequenceserver.rb]: Use `Hash#fetch` instead of `[]`. Rescue `KeyError`, and re-raise it as 400 Bad Request.

**Comment by yeban** [comment]
Need to add logging for debug mode.

**Comment by yeban** [comment]
Safari add `.txt` extension to XML download. Will need to use appropriate mime type.

**Comment by yeban** [comment]
Need to remove `[]` from around the links. Should use `|` instead of `/` as the separator like everywhere else in the HTML report.

**Comment by yeban** [comment]
Should inform users what columns standard TSV will contain. Via BS tooltip, perhaps.

**Pull request merged**